### ISSUE TITLE
skydns fails to start when podless service exists

### ISSUE THREAD

**Issue by fluxrad** [kubernetes/kubernetes#25573]
It appears SkyDNS will fail to start if a service exists that exclusively has failed pods behind it. Both SkyDNS and Kube2Sky enter a failed state causing all kube-dns pods to eventually enter a CrashLoopBackoff state. Steps to reproduce:

1. Create a kube-dns add-on service (in our case 5 replicas).
1. Create a Deployment in the `kube-system` namespace. In our case, Prometheus with a single replica.
1. Create a service in the `kube-system` namespace for the Deployment
1. Have the pod backing the service enter a failed state, so the service has no running pods behind it.
1. Restart kube-dns pods. They should enter a crash loop. 

The symptom appears to be a crash loop and the following error message:

```
$ kubectl logs kube-dns-v11-eek1r -c skydns --namespace=kube-system
2016/05/13 15:06:23 skydns: falling back to default configuration, could not read from etcd: 100: Key not found (/skydns/config) [2037]
2016/05/13 15:06:23 skydns: bad port for prometheus: tcp://10.100.102.66:9090
```

Deleting the pod-less service immediately fixes the issue.

**Comment by thesuperzapper** [comment]
I think I have this problem also, did you manage to get a working DNS? If, so I could use a hint.

**Comment by thockin** [comment]
@ArtfulCoder @girishkalele

**Comment by girishkalele** [comment]
@fluxrad 

Could you provide more details of your setup ?
Which version are you running ?
I think the DNS addon is now part of the standard installed services in 1.2.

**Comment by fluxrad** [comment]
OS: CoreOS (stable)
Docker version: 1.9.1
Kubernetes version: 1.2.3
Kube DNS add-on version: v11

Kubernetes installed via CloudFormation bootstrap. Kubelet is a daemon, other services started via Hyperkube in Docker. Cribbed from an earlier version of CoreOS' kube-aws tool.

DNS is configured based on the cluster add-on specified in [kubernetes/addons/dns](https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/dns/skydns-rc.yaml.in) (latest master; v11)

**Comment by girishkalele** [comment]
@thockin @fluxrad 

The exact failure is due to the failure to atoi parse the string "tcp://10.100.102.66:9090" in skydns code.
This is the value of the PROMETHEUS_PORT environment variable which triggers the "export metrics to prometheus" feature in skydns (I know, I know, activating a feature just due to the presence of an environment variable is bad practice).

Because your service is named prometheus and runs in the kube-system namespace, all pods run in that namespace will have a variety of environment variables set including PROMETHEUS_PORT. When you restart the dns addon pod after creating the prometheus service, skydns triggers on this and panics.

$ execp prome env | grep PROMETHEUS
PROMETHEUS_SERVICE_HOST=10.0.68.8
PROMETHEUS_PORT_5201_TCP=tcp://10.0.68.8:5201
PROMETHEUS_PORT_5201_TCP_PORT=5201
PROMETHEUS_SERVICE_PORT=5201
PROMETHEUS_PORT=tcp://10.0.68.8:5201   **<<<<<<<<< the atoi() failure comes from trying to parse this**
PROMETHEUS_PORT_5201_TCP_PROTO=tcp
PROMETHEUS_SERVICE_PORT_IPERF=5201
PROMETHEUS_PORT_5201_TCP_ADDR=10.0.68.8

>>// Metrics registers the DNS metrics to Prometheus, and starts the internal metrics
>>// server if the environment variable PROMETHEUS_PORT is set.
>>func Metrics() error {
>>	// We do this in a function instead of using var + init(), because we want to
>>	// able to set Namespace and/or Subsystem.
>>	if Port == "" {
>>		return nil
>>	}
>>
>>	_, err := strconv.Atoi(Port)
>>	if err != nil {
>>		fmt.Errorf("bad port for prometheus: %s", Port)
>>	}

**Comment by girishkalele** [comment]
Suggested workarounds for now:

Naming your service something other than prometheus or, less suitable, not running in the kube-system namespace.

**Comment by girishkalele** [comment]
Cannot reproduce when retested with the new kubedns implementation - the presence of the environment variable does not affect the kubedns pods now.

**Comment by ntquyen** [comment]
Ran into this issue in kubernetes 1.2.0, kube-dns-v11, took us 2 hours to figure it out...

**Comment by girishkalele** [comment]
@ntquyen 

I am wondering if there is documentation out there configuring a "prometheus" service in the kube-system namespace. Can you point me to the docs you used to setup your prometheus service ?

**Comment by ntquyen** [comment]
I followed this link https://coreos.com/blog/prometheus-and-kubernetes-up-and-running.html. It doesn't say about running in `kube-system`, I decided myself, old habit, monitoring tools all go into `kube-system`...

**Comment by girishkalele** [comment]
The new kube-dns code in 1.3+ does not read or depend on this PROMETHEUS environment variable. Marking as Closed for 1.3.
